### PR TITLE
Add new line protocol parser and serializer, influxdb output

### LINKED ISSUE #3291: Metrc Interface and ReplaceFieldValue function
### Feature Request
`Metric` interface contains `AddField`and `RemoveField` functions.
It will be good if this interface has function `ReplaceField` for replace field value.

### LINKED ISSUE #3452: InfluxDB output plugin should sort tags
Sending unsorted tags to InfluxDB generates additional work, including allocating a new slice for each point that is parsed and subsequently sorting the tag key / values.

Benchmarks for parsing 5000 points with 10 sorted vs unsorted tags:

```
± go test -bench=_5000 -run=- ./models
goos: darwin
goarch: amd64
pkg: github.com/influxdata/influxdb/models
BenchmarkParsePointsTagsUnsorted10_5000-8   	     100	  10185633 ns/op	  70.69 MB/s	 1921936 B/op	   10001 allocs/op
BenchmarkParsePointsTagsSorted10_5000-8     	     200	   6268681 ns/op	 114.86 MB/s	 1281920 B/op	    5001 allocs/op
```

### LINKED ISSUE #3631: Invalid number error when parsing line protocol with dos line endings
### Relevant telegraf.conf:
```toml
[[inputs.exec]]
   commands = ['sh -c "cat ~/telegraf.txt"']
   timeout = "25s"
   data_format = "influx"
```

### System info:
```
$ ./telegraf --version
Telegraf v1.6.0~06c21fb9 (git: master 06c21fb9)
```

### Steps to reproduce:
```
$ cat ~/telegraf.txt
dt,entity=SERVICE-F08D806E7EF685CF timeseriesId="com.dynatrace.builtin:service.responsetime",metric=56167.75
dt,entity=SERVICE-F08D806E7EF685CF timeseriesId="com.dynatrace.builtin:service.responsetime",metric=65166.0
$ file ~/telegraf.txt
/home/dbn/telegraf.txt: ASCII text, with CRLF line terminatorst
$ telegraf --test
2018/01/02 10:54:07 I! Using config file: /home/dbn/.telegraf/telegraf.conf
* Plugin: inputs.exec, Collection 1
2018-01-02T18:54:07Z E! Error in plugin [inputs.exec]:  invalid number invalid number
```

### Expected behavior:
```
$ telegraf --test
2018/01/02 11:00:57 I! Using config file: /home/dbn/.telegraf/telegraf.conf
* Plugin: inputs.exec, Collection 1
> dt,entity=SERVICE-F08D806E7EF685CF,host=loaner timeseriesId="com.dynatrace.builtin:service.responsetime",metric=56167.75 1514919658000000000
> dt,entity=SERVICE-F08D806E7EF685CF,host=loaner timeseriesId="com.dynatrace.builtin:service.responsetime",metric=65166 1514919658000000000
```

### Actual behavior:

```
2018-01-02T18:54:07Z E! Error in plugin [inputs.exec]:  invalid number invalid number
```

### Additional info:

Changing the file to unix line endings results in correct parsing.

### LINKED ISSUE #3648: Telegraf no more working when changing ip adress in DNS name of InfluxDB server
### System info:

Telegraf v1.5.0 on CentOS 6.5

### Steps to reproduce:

1. Start Telegraf pointing to influxdb.example.com (a CNAME to gateway.example.com, which is a reverse-proxy to the real InfluxDB server)
2. Telegraf works fine.
3. Change IP address of gateway.example.com
4. Wait several minutes/hours so that DNS servers and DNS cache take into account the new IP address
5. curl -v http://influxdb.example.com/ping works fine on the server with the Telegraf agent
6. Telegraf does not work anymore except if you restart it

Seems that IP address of the InfluxDB server is cached by the Telegraf agent (or perhaps a Go library) and never updated.

### PR DESCRIPTION
**Pull request by danielnelson**
In this pull request I introduce a new line protocol parser and serializer, as well as an updated influxdb output that uses the serializer.  Some minor changes are made to the `telegraf.Metric` interface as well.

Line protocol is no longer the internal represention of the `metrics` type, the data is stored in a fully parsed and type converted manor.  Despite this in my testing there is an performance improvement across both going from line protocol to line protocol, as well as with Telegraf generated metrics going to line protocol, the main reason for this is the ragel based parser.  Additionally, the sort order of tags is fixed, which should help on the InfluxDB side.

[Ragel](https://www.colm.net/open-source/ragel/) is used to compile the parser state machine, the generated file is committed so that there is no build requirement on having ragel.  In the future I would love to have this be done at build time so that no generated files are committed.

closes #3452
closes #3631
closes #3648
closes #3291

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

**Comment by danielnelson** [comment]
@phemmer This is a big one, so I won't ask you to do a full review, but I'd grateful for any feedback you have time to provide.  I'm excited about the prospect of converting other parsers to use ragel in the future (statsd parsing comes to mind).

**Comment by danielnelson** [comment]
@ragzilla has been working on a follow up patch adding uint support: https://github.com/ragzilla/telegraf/commit/3064edc11c96fff32b8de1bc7700ed086f41b0f7#commitcomment-28296437

We have to be a little careful as uint support is not generally available in InfluxDB and any field type changes results in an error.

File: internal/models/makemetric.go [hunk 1]
```diff
@@ -78,84 +76,6 @@ func makemetric(
...
-			if val < uint64(9223372036854775808) {
-				fields[k] = int64(val)
-			} else {
-				fields[k] = int64(9223372036854775807)
-			}
-			continue
-		case float32:
-			fields[k] = float64(val)
-			continue
-		case float64:
```
Comment (buddyledungarees) [inline internal/models/makemetric.go]: Sorry for opening an old issue but this check was blocking a lot of garbage from being attempted to be written to our output. Is there a different place that this same kind of filtering can be done?

Comment (danielnelson) [reply internal/models/makemetric.go]: Can you describe what you are seeing now in a new issue?

**Comment by stevenh** [comment]
I've just identified this as the cause of a major performance regression as detailed here:
https://github.com/influxdata/telegraf/issues/7070

**Pull request merged**